### DISCUSSION TITLE
osqp used with an MCU(STM32 F405/F407)

### DISCUSSION THREAD

**Discussion by andreaskruger** [osqp/osqp#614] [C Interface] [answered]
Is there any documentation on the use of OSQP on an STM32 F407? Trying to use the OSQP on my MCU but cant seem to find any information how or what to add to my MCU project.

Any suggestions are welcome

**Comment by imciner2** [comment, marked as answer]
Hi @andreaskruger, sorry for not replying earlier. OSQP should be capable of being used on these micro controllers, but the exact files you will need depend on the version of OSQP you are using. Which version are you trying?

For OSQP 0.6, you will need to add all the source files from the src and include directories to your project, and also set the `OSQP_EMBEDDED_MODE` define to either 1 or 2 (depending on if you want to allow vector updates only, or both vector and matrix updates). You can then use the code generation facilities in the Python or Matlab interfaces to generate the matrices/vectors and the solver that you can then deploy on your MCU.

**Reply by guangyad** [reply]
Hi @imciner2,

I’m following up on this closed issue as I’ve run into the same porting challenge across multiple versions of OSQP (including but not limited to 0.6.0 and 1.0.0) when trying to deploy it on an STM32F407 MCU (Cortex-M4F with FPU, using Keil MDK for pure C embedded development). The guidance you provided for version 0.6 unfortunately no longer works with the current codebase structure, which has changed dramatically since that release.

When I tried to follow the original instructions — adding all source files from the `src` directory and headers from the top-level `include` directory to my STM32 project — I got a cascade of missing header errors and unresolved dependencies, regardless of the version I attempted. Many required header files are not housed in the main `include` folder, but are spread across other subdirectories in the repository, and it’s nearly impossible to untangle which files are strictly required for a minimal embedded deployment, and which are for desktop/other use cases.

I’m hoping to get updated guidance for deploying OSQP on STM32 Cortex-M MCUs:
1.  What are the mandatory source files and header directories that must be included for a minimal, functional port?
2.  What are the required macro definitions (such as `OSQP_EMBEDDED_MODE`, memory configuration, floating-point settings, etc.) that need to be set correctly for this embedded use case?
3.  Is there any official minimal porting guide, or example project, for OSQP targeting Cortex-M series MCUs?

This updated guidance would be incredibly helpful not just for me, but for the many embedded developers working with STM32 MCUs who want to use OSQP. Any clarification or help you can provide would be greatly appreciated.

Thank you!

**Comment by andreaskruger** [comment]
Hi @imciner2, no worries. There were no problems at all implementing OSQP on a micro controller such as an STM32f407g and it runs very smoothly. 

Thanks for the information though.

Andreas

**Reply by YL1142** [reply]
Hi, I'm very interested in this issue, and I also want to implement OSQP  on a micro controller such as an STM32f407g, but I'm a bit unsure of where to start. If you have implemented this functionality, I would like to ask you for guidance.
